\documentclass[]{article}

%opening
\title{}
\author{O. Uriarte, D. Kyriakis, T. Heurtaux, K. Grzyb, R. Halder, M. Buttini, A. Skupin, M. Mittelbronn& A. Michelucci}

\begin{document}

\maketitle

\begin{abstract}

\end{abstract}

\section{Methods}
\subsection{Single cell RNA-seq data analysis}
We filtered out the low quality cells and genes separately in each data set. We defined cells as low-quality, based on three criteria for each cell.  The number of the genes that expressed is more than 200 and 2 median-absolute- deviations (MADs) above the median, the total number of counts is 3 MADs above or below the median and the percentage of counts to mitochondrial genes is 1.5 median-absolute- deviations (MADs) above the median. Cells failing at least one criterion were considered as low quality cells and filtered out from further analysis (Sup. Fig1). Similar to the cell filtering, we filtered out the low quality genes that been expressed in less than 10 cells in the data.\newline
We used \texttt{Seurat v3.1} for the analysis of the gene expression matrix. We merged the gene expression matrices from striatum and midbrain. After the filtering, the count data of 1036 cells and 16648 genes was used for the downstream analysis. To identify the different cell types, we clustered the cells and visualized clusters using a t-distributed stochastic neighbor embedding (t-SNE) plot (Fig. \ref{fig:1}, b). The resolution for the Louvain clustering, selected based on the silhouette score performance of different resolutions. This revealed eight main clusters.\newline
In order to recover the cell type identity of each cluster we performed one unsupervised and one supervised workflow.  In the former, we performed a differential expression analysis between the clusters. The genes that popped up as differentially expressed, were searched in literature and we linked them with specific cell types (Fig. \ref{fig:1}, d). In the latter one, we used known cell specific markers (Fig. \ref{fig:1}, e).  These two workflows let us assign each cluster to cell types (Fig. \ref{fig:1}, c,f). 
Since we focus here on the microglia, we subset the data and keep only the cells that were identified as microglia. We re-project and cluster these cells. This revealed four different clusters (Fig. \ref{fig:2}, a). The minor cluster (cluster 4) seems to express also oligodendrocyte markers so we filtered out the four cluster. Then we performed differential expression analysis between the brain regions (Fig. \ref{fig:2}, b).


\begin{figure}[]
	\caption{ Two-dimensional t-SNE plot of 1036 individual cells at two different regions of mouse brain. Each dot represents a single cell. \textbf{a)} The colors correspond to different regions of the mouse brain. \textbf{b)} t-SNE plot where the colors correspond to eight different clusters. \textbf{c)} The colors correspond to different cell types, which were assigned based on specific cell markers. \textbf{d)} Heat map of the top 15 differentially expressed genes for each cluster. \textbf{e)} Heat map using specific cell type markers. \textbf{f)} Bar plot of cell type markers of each cell, colored based on the different cell types. \textbf{g)} Bar plot of with the number of cells from the different brain regions per clusters.}
	\label{fig:1}
\end{figure}

\begin{figure}[]
	\caption{ Heatmap of the differentially expressed genes. \textbf{a)} Differentially expression analysis between the microglia subpopulations. \textbf{b)} Differentially expression analysis between the striatum and midbrain.}
	\label{fig:2}
\end{figure}




\begin{figure}[]
	\caption{ Depiction of criteria for cell quality assessment and filtering.  \textbf{(a-c)} Distributions of relevant quality criteria for single cells analyzed. The number of counts per cell, number of detected features per cell and percentage of counts to mitochondrial features are shown (in this order). \textbf{(d-e)}  The t-SNE projection, after the low quality cell filtering, of the number of genes that expressed per cell and total counts per cell. }
	\label{fig:3}
\end{figure}

\end{document}


